File: step_13_binary_trees/13.2_medium/IB101_sum_root_to_leaf_numbers/solution.py
"""
Sum Root to Leaf Numbers (LeetCode #129)

Each root-to-leaf path forms a number. Return the total sum of all numbers.
"""
from typing import Optional
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# ============================================================
# APPROACH 1: BRUTE FORCE -- Collect all paths as strings
# Time: O(N * H)  |  Space: O(N * H)
# ============================================================
def brute_force(root: Optional[TreeNode]) -> int:
    if not root:
        return 0

    paths = []

    def collect(node, path):
        if not node:
            return
        path += str(node.val)
        if not node.left and not node.right:
            paths.append(path)
            return
        collect(node.left, path)
        collect(node.right, path)

    collect(root, "")
    return sum(int(p) for p in paths)


# ============================================================
# APPROACH 2: OPTIMAL -- Top-down DFS with running number
# Time: O(N)  |  Space: O(H)
# ============================================================
def optimal(root: Optional[TreeNode]) -> int:
    def dfs(node, current_num):
        if not node:
            return 0
        current_num = current_num * 10 + node.val
        # Leaf: return the complete number
        if not node.left and not node.right:
            return current_num
        # Internal: sum of both subtrees
        return dfs(node.left, current_num) + dfs(node.right, current_num)

    return dfs(root, 0)


# ============================================================
# APPROACH 3: BEST -- Iterative BFS with running numbers
# Time: O(N)  |  Space: O(W)
# ============================================================
def best(root: Optional[TreeNode]) -> int:
    if not root:
        return 0

    total = 0
    queue = deque([(root, root.val)])

    while queue:
        node, num = queue.popleft()

        if not node.left and not node.right:
            total += num
            continue

        if node.left:
            queue.append((node.left, num * 10 + node.left.val))
        if node.right:
            queue.append((node.right, num * 10 + node.right.val))

    return total


# ============================================================
# MAIN
# ============================================================
if __name__ == "__main__":
    #      4
    #     / \
    #    9   0
    #   / \
    #  5   1
    root = TreeNode(4)
    root.left = TreeNode(9)
    root.right = TreeNode(0)
    root.left.left = TreeNode(5)
    root.left.right = TreeNode(1)

    print("=== Sum Root to Leaf Numbers ===")
    print(f"Brute:   {brute_force(root)}")   # 1026
    print(f"Optimal: {optimal(root)}")        # 1026
    print(f"Best:    {best(root)}")           # 1026
    # Paths: 495 + 491 + 40 = 1026

File: step_13_binary_trees/13.2_medium/IB101_sum_root_to_leaf_numbers/Solution.java
import java.util.*;

/**
 * Sum Root to Leaf Numbers (LeetCode #129)
 *
 * Each root-to-leaf path forms a number. Return the total sum of all numbers.
 */
public class Solution {

    static class TreeNode {
        int val;
        TreeNode left, right;
        TreeNode(int val) { this.val = val; }
    }

    // ============================================================
    // APPROACH 1: BRUTE FORCE -- Collect all paths as strings
    // Time: O(N * H)  |  Space: O(N * H)
    // ============================================================
    public static int bruteForce(TreeNode root) {
        if (root == null) return 0;

        List<String> paths = new ArrayList<>();
        collectPaths(root, "", paths);

        int total = 0;
        for (String path : paths) {
            total += Integer.parseInt(path);
        }
        return total;
    }

    private static void collectPaths(TreeNode node, String path, List<String> paths) {
        if (node == null) return;
        path += node.val;
        if (node.left == null && node.right == null) {
            paths.add(path);
            return;
        }
        collectPaths(node.left, path, paths);
        collectPaths(node.right, path, paths);
    }

    // ============================================================
    // APPROACH 2: OPTIMAL -- Top-down DFS with running number
    // Time: O(N)  |  Space: O(H)
    // ============================================================
    public static int optimal(TreeNode root) {
        return dfs(root, 0);
    }

    private static int dfs(TreeNode node, int currentNum) {
        if (node == null) return 0;

        currentNum = currentNum * 10 + node.val;

        // Leaf node: return the complete number
        if (node.left == null && node.right == null) {
            return currentNum;
        }

        // Internal node: sum of left and right subtree numbers
        return dfs(node.left, currentNum) + dfs(node.right, currentNum);
    }

    // ============================================================
    // APPROACH 3: BEST -- Iterative BFS with running numbers
    // Time: O(N)  |  Space: O(W)
    // ============================================================
    public static int best(TreeNode root) {
        if (root == null) return 0;

        int totalSum = 0;
        Queue<TreeNode> nodeQueue = new LinkedList<>();
        Queue<Integer> numQueue = new LinkedList<>();
        nodeQueue.offer(root);
        numQueue.offer(root.val);

        while (!nodeQueue.isEmpty()) {
            TreeNode node = nodeQueue.poll();
            int num = numQueue.poll();

            if (node.left == null && node.right == null) {
                totalSum += num;
                continue;
            }

            if (node.left != null) {
                nodeQueue.offer(node.left);
                numQueue.offer(num * 10 + node.left.val);
            }
            if (node.right != null) {
                nodeQueue.offer(node.right);
                numQueue.offer(num * 10 + node.right.val);
            }
        }

        return totalSum;
    }

    // ============================================================
    // MAIN
    // ============================================================
    public static void main(String[] args) {
        //      4
        //     / \
        //    9   0
        //   / \
        //  5   1
        TreeNode root = new TreeNode(4);
        root.left = new TreeNode(9);
        root.right = new TreeNode(0);
        root.left.left = new TreeNode(5);
        root.left.right = new TreeNode(1);

        System.out.println("=== Sum Root to Leaf Numbers ===");
        System.out.println("Brute:   " + bruteForce(root));  // 1026
        System.out.println("Optimal: " + optimal(root));      // 1026
        System.out.println("Best:    " + best(root));          // 1026
        // Paths: 495 + 491 + 40 = 1026
    }
}
